diffsims/generators/diffraction_generator.py

```diff
@@ -149,21 +297,50 @@ def calculate_ed_data(
         # excitation error and store the magnitude of their excitation error.
         r_sphere = 1 / wavelength
         r_spot = np.sqrt(np.sum(np.square(cartesian_coordinates[:, :2]), axis=1))
-        z_sphere = -np.sqrt(r_sphere ** 2 - r_spot ** 2) + r_sphere
-        excitation_error = np.absolute(z_sphere - cartesian_coordinates[:, 2])
-        intersection = excitation_error < max_excitation_error
+        z_spot = cartesian_coordinates[:, 2]
+
+        if self.precession_angle > 0 and not self.approximate_precession:
+            # We find the average excitation error - this step can be
+            # quite expensive
+            excitation_error = _average_excitation_error_precession(
+                    z_spot,
+                    r_spot,
+                    wavelength,
+                    self.precession_angle,
+                    )
+        else:
+            z_sphere = -np.sqrt(r_sphere ** 2 - r_spot ** 2) + r_sphere
+            excitation_error = z_sphere - z_spot
         # Mask parameters corresponding to excited reflections.
+        intersection = np.abs(excitation_error) < max_excitation_error
         intersection_coordinates = cartesian_coordinates[intersection]
-        g_indices = spot_indices[intersection]
         excitation_error = excitation_error[intersection]
+        r_spot = r_spot[intersection]
+        g_indices = spot_indices[intersection]
         g_hkls = spot_distances[intersection]
-
-        if shape_factor_model is not None:
-            shape_factor = shape_factor_model(
-                excitation_error, max_excitation_error, **kwargs
-            )
+        # take into consideration rel-rods
+        if self.precession_angle > 0 and not self.approximate_precession:
```

Comment:
picky, but I would probably find these easier to read if they were nested:
```python
if self.precession_angle > 0:
        if not self.approximate_precession: 
        
        else:
else:
```

Reply:
There is no nested else though. Only if one has precession AND one wants a full calculation does the mean excitation error need to be calculated. In all other cases we just calculate it the simple way. I'll implement it as you say and one can return to it later.

Reply:
I thought one could refactor the approximation inside?
```python
if self.precession_angle > 0:
        if not self.approximate_precession: 
        
        elif self.approximate_precession:

elif self.precession_angle == 0:
```
(I've made everything `elif` in this comment to aid clarity)

Reply:
Unfortunately I've changed it back to what it was originally, if you find a better way feel free to edit. Reason is that in the `elif self.approximate_precession` i would have the same code as in the `elif self.precession_angle == 0`, which would be unnecessary duplication. This only influences how the `excitation_error` is calculated and which spots need to be thrown out in the first stage. If there is precession but we are content with the approximation, then we calculate the excitation error in the same way as if there was no precession. If the precession is 0 but we set approximate precession to True, we also want to execute the else block.